resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/resources/ZipFileTransformer.java

```diff
@@ -0,0 +1,113 @@
+ * https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors: Vector Informatik GmbH - initial API and implementation
+ *******************************************************************************/
+
+package org.eclipse.core.resources;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.URI;
+import java.net.URISyntaxException;
+import org.eclipse.core.filesystem.EFS;
+import org.eclipse.core.filesystem.IFileStore;
+import org.eclipse.core.filesystem.URIUtil;
+import org.eclipse.core.filesystem.ZipFileUtil;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IPath;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Status;
+import org.eclipse.core.runtime.SubMonitor;
+
+/**
+ * Utility class for opening and closing zip files.
+ *
+ * @since 3.21
+ */
+public class ZipFileTransformer {
+
+	/**
+	 * Closes an opened zip file represented as a linked folder in the workspace.
+	 * After closing, the zip file in its file state is shown in the workspace.
+	 *
+	 * This method can only be called when the zip file is local. Otherwise a
+	 * CoreException is thrown.
+	 *
+	 * @param folder The folder representing the zip file to close.
+	 *
+	 */
+	public static void closeZipFile(IFolder folder) throws URISyntaxException, CoreException {
+		URI zipURI = new URI(folder.getLocationURI().getQuery());
+		IFileStore parentStore = EFS.getStore(folder.getParent().getLocationURI());
+		URI childURI = parentStore.getChild(folder.getName()).toURI();
+		if (URIUtil.equals(zipURI, childURI)) {
+			folder.delete(IResource.CLOSE_ZIP_FILE, null);
+			folder.getProject().refreshLocal(IResource.DEPTH_INFINITE, null);
+		} else {
+			throw new CoreException(new Status(IStatus.ERROR, ResourcesPlugin.PI_RESOURCES,
+					"Closing of Zip File " + folder.getName() //$NON-NLS-1$
+							+ " failed because the Zip File is not local.")); //$NON-NLS-1$
+		}
+	}
+
+	/**
+	 * Opens a zip file represented by a file into a linked folder. The zip file
+	 * will not be extracted in this process. In the opened state the linked folder
+	 * allows reading and manipulating the children of the zip file in the workspace
+	 * and on the filesystem. If the folder has no children after opening, then it
+	 * is closed immediately after. In this case it could be that there should be no
+	 * children, so there is no need for opening or an error occured that prevented
+	 * the children from being loaded.
+	 *
+	 * This method prevents opening linked zip files. zip files must be local to be
+	 * opened. Otherwise a CoreException is thrown.
+	 *
+	 * @param file    The file representing the zip file to open.
+	 * @param monitor monitor indicating the completion progress
+	 *
+	 */
+	public static void openZipFile(IFile file, IProgressMonitor monitor, boolean backgroundRefresh)
```

Comment:
opening a zip file only takes some milliseconds, not worth a ProgressMonitor.
"backgroundRefresh" is not documented.

Reply:
I added documentation and removed the monitor in commit 91f9a51af1e29228cd3014a69be650a69518fe23. We had performance issues and used a ProgressDialog in UI and the ProgressMonitor here because of the performance issues. But now the issues are solved so I removed these changes.